### DISCUSSION TITLE
Istio CNI - memory spikes and OOM errors

### DISCUSSION THREAD

**Discussion by AleksanderBrzozowski** [istio/istio#48922] [Q&A]
Hi, we have a EKS Kubernetes cluster in AWS with Istio mesh, version 1.17.8. 

We use Istio CNI to configure networking, and we observe memory spikes every ~ 50 minutes:

![image](https://github.com/istio/istio/assets/15987499/73805468-4cdf-41c9-abab-dae3e1ab2348)

We figured out that it is caused by service account token refresh, but what we don't understand is why memory increases suddenly? These memory spikes sometimes cause OOM errors. The same memory spike we observe during startup.

We tried upgrading Istio to 1.19.6, but we observed the same behavior after upgrade. Furthermore, setting `resources.memory.limits` of the Istio CNI deployment to `100Mi` causes it to fail instantly with OOM.

Here are Istio CNI logs when service account is refreshed (from version 1.19.6): 

```text
2024-01-23T13:08:42.748082Z    info    file modified: /var/run/secrets/kubernetes.io/serviceaccount/..REDACTED
2024-01-23T13:08:42.755225Z    info    install    Detect changes to the CNI configuration and binaries, attempt reinstalling...
2024-01-23T13:08:44.218934Z    info    install    Copied istio-cni to /host/opt/cni/bin.
2024-01-23T13:08:44.218976Z    info    install    Directory /host/secondary-bin-dir is not writable, skipping.
2024-01-23T13:08:44.220033Z    info    install    write kubeconfig file /host/etc/cni/net.d/ZZZ-istio-cni-kubeconfig with:
apiVersion: v1
clusters:
- cluster:
    certificate-authority: REDACTED
    certificate-authority-data: REDACTED
    server: REDACTED
  name: local
contexts:
- context:
    cluster: local
    user: istio-cni
  name: istio-cni-context
current-context: istio-cni-context
kind: Config
preferences: {}
users:
- name: istio-cni
  user:
    token: REDACTED

2024-01-23T13:08:44.220058Z    info    install    Using CNI config template from CNI_NETWORK_CONFIG environment variable.
2024-01-23T13:08:44.220075Z    info    install    CNI config: {
  "cniVersion": "0.3.1",
  "name": "istio-cni",
  "type": "istio-cni",
  "log_level": "debug",
  "log_uds_address": "/var/run/istio-cni/log.sock",

  "kubernetes": {
      "kubeconfig": "/etc/cni/net.d/ZZZ-istio-cni-kubeconfig",
      "cni_bin_dir": "/opt/cni/bin",
      "exclude_namespaces": [ "istio-system", "kube-system" ]
  }
}
2024-01-23T13:08:44.220386Z    info    install    CNI config file /host/etc/cni/net.d/10-aws.conflist exists. Proceeding.
2024-01-23T13:08:44.220880Z    info    install    Created CNI config /host/etc/cni/net.d/10-aws.conflist
2024-01-23T13:08:44.220899Z    info    install    CNI configuration and binaries reinstalled.
2024-01-23T13:08:44.220944Z    info    file watcher skipping watch on non-existent path: /host/secondary-bin-dir/istio-cni
```

Resources:

```yaml
        resources:
          limits:
            cpu: 200m
            memory: 256Mi
          requests:
            cpu: 50m
            memory: 100Mi
```

Container settings:

```yaml
     containers:
      - args:
        - --log_output_level=default:info,cni:info
        command:
        - install-cni
        env:
        - name: CNI_NETWORK_CONFIG
          valueFrom:
            configMapKeyRef:
              key: cni_network_config
              name: istio-cni-config
        - name: CNI_NET_DIR
          value: /etc/cni/net.d
        - name: CHAINED_CNI_PLUGIN
          value: "true"
        - name: REPAIR_ENABLED
          value: "true"
        - name: REPAIR_NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        - name: REPAIR_LABEL_PODS
          value: "true"
        - name: REPAIR_DELETE_PODS
          value: "true"
        - name: REPAIR_RUN_AS_DAEMON
          value: "true"
        - name: REPAIR_SIDECAR_ANNOTATION
          value: sidecar.istio.io/status
        - name: REPAIR_INIT_CONTAINER_NAME
          value: istio-validation
        - name: REPAIR_BROKEN_POD_LABEL_KEY
          value: cni.istio.io/uninitialized
        - name: REPAIR_BROKEN_POD_LABEL_VALUE
          value: "true"
```

The question is - should we just increase `memory.limits` of Istio CNI pods to avoid OOM errors, or maybe there is something wrong with our setup?

**Comment by howardjohn** [comment]
there might be room for improvement here on the memory usage but increasing the limit seems like an obvious short term step to make an improvement

**Reply by AleksanderBrzozowski** [reply]
@howardjohn Is Istio CNI a subject of performance test? I mean, there is a page with [performance and scalability of Istio](https://istio.io/latest/docs/ops/deployment/performance-and-scalability/), where you measure data plane and control plane performance, but is there a similar experiment run for Istio CNI? If it's not, then do you know what memory usage we should expect from Istio CNI? 🙂

**Reply by howardjohn** [reply]
I don't think we regularly run performance tests here. It should use very little memory, it does extremely little work.

**Reply by AleksanderBrzozowski** [reply]
Any advice on how we can troubleshoot memory usage in our case?
It uses very little memory but only after installation is completed, and when it is in idle state.

**Reply by howardjohn** [reply]
The normal way is to use pprof but looks like its not enabled on CNI. Best would probably be to make a small tweak with pprof and try to reproduce it. It would also be good to see if we see this in other environments or only yours

**Reply by AleksanderBrzozowski** [reply]
@howardjohn 

Well, I don't think that I can set up pprof without making any changes to Istio CNI? As I understand, we need to add "instrumentation code" to the Istio CNI first, and enable pprof endpoints, right?

I am going to enable debugging logs, maybe this will help.

**Reply by howardjohn** [reply]
Yeah I was mostly thinking an istio dev would do that and try to reproduce (or you become an Istio dev :-) https://github.com/istio/istio/wiki/ has getting started info to try it yourself if you are interested)

**Reply by AleksanderBrzozowski** [reply]
Yeah, thanks for sending the docs how to get started 👍 

I forked the repo, but I am not sure when I will be able to start adding pprof support for Istio CNI. Therefore, I've created a feature request -> https://github.com/istio/istio/issues/49053

If no one picks it up, I will try to do it by myself.

**Reply by howardjohn** [reply]
I am fairly sure https://github.com/istio/istio/pull/50393 will fix this. Nice find!
